python/ray/serve/tests/unit/test_application_state.py

```diff
@@ -600,20 +607,83 @@ def test_deploy_through_config_fail(check_obj_ref_ready_nowait):
     app_state.update()
     assert app_state.status == ApplicationStatus.DEPLOYING
 
-    # Object ref is ready, and the task has called apply_deployment_args
+    # Object ref is ready, and the task has called deploy_app
     check_obj_ref_ready_nowait.return_value = True
     app_state.update()
     assert app_state.status == ApplicationStatus.DEPLOY_FAILED
     assert "failed" in app_state._status_msg or "error" in app_state._status_msg
 
 
+@patch(
+    "ray.serve._private.application_state.get_app_code_version",
+    Mock(return_value="123"),
+)
+@patch("ray.serve._private.application_state.build_serve_application", Mock())
+@patch("ray.get", Mock(return_value=([deployment_params("a", "/old", "/docs")], None)))
+@patch("ray.serve._private.application_state.check_obj_ref_ready_nowait")
+def test_apply_app_configs_deletes_existing(check_obj_ref_ready_nowait):
+    """Test that apply_app_configs deletes existing apps that aren't in the new list.
+
+    This should *not* apply to apps that were deployed via `deploy_app` (which is
+    an imperative API).
+    """
+    kv_store = MockKVStore()
+    deployment_state_manager = MockDeploymentStateManager(kv_store)
+    app_state_manager = ApplicationStateManager(
+        deployment_state_manager, MockEndpointState(), kv_store
+    )
+
+    # Deploy an app via `deploy_app` - should not be affected.
+    a_id = DeploymentID(name="a", app_name="imperative_app")
+    app_state_manager.deploy_app("imperative_app", [deployment_params("a", "/hi")])
+    imperative_app_state = app_state_manager._application_states["imperative_app"]
+    assert imperative_app_state.api_type == APIType.IMPERATIVE
+    assert imperative_app_state.status == ApplicationStatus.DEPLOYING
+
+    imperative_app_state.update()
+    deployment_state_manager.set_deployment_healthy(a_id)
+    imperative_app_state.update()
+    assert imperative_app_state.status == ApplicationStatus.RUNNING
+
+    # Now deploy an initial version of the config with app 1 and app 2.
+    app1_config = ServeApplicationSchema(
+        name="app1", import_path="fa.ke", route_prefix="/1"
+    )
+    app2_config = ServeApplicationSchema(
+        name="app2", import_path="fa.ke", route_prefix="/2"
+    )
+    app_state_manager.apply_app_configs([app1_config, app2_config])
+    app1_state = app_state_manager._application_states["app1"]
+    assert app1_state.api_type == APIType.DECLARATIVE
+    app2_state = app_state_manager._application_states["app2"]
+    assert app2_state.api_type == APIType.DECLARATIVE
+    app1_state.update()
+    app2_state.update()
```

Comment:
nit: why not just application_state_manager.update()?

Reply:
I was just copying the existing tests and was a bit lost. let me try this insteaad

Reply:
ok somehow that is not equivalent :(

Reply:
leaving it as-is ... can you fix these tests properly sometime soon pls?